de.fu_berlin.inf.dpp.whiteboard/frontend/src/services/javabridge.js

```diff
@@ -0,0 +1,145 @@
+      this.canvas.renderAll();
+    };
+    //sets the state of the whiteboard at the begining of the session using data from the Java Model
+    window.setState = (state) => {
+      //convert recursive data structure to linear structure
+      let records = flatten(state);
+      //ignore root element
+      records.shift();
+      //reset
+      this.viewManager.clear();
+      records.forEach(record => this.createNew(record));
+      this.canvas.renderAll();
+    };
+  }
+
+  /**
+   * executes a given record
+   * @param {Object} record
+   */
+  executeRecord(record) {
+    let action = record.type.toLowerCase();
+    let props = record.valuePairs;
+    switch (action) {
+      case SXE.ACTIONS.NEW:
+        return this.createNew(props);
+      case SXE.ACTIONS.SET:
+        return this.set(props);
+      //the current java implementation of the SXE protocol uses "SET":visible=false instead of "REMOVE"
+      //because its faster (on java level), that why it is not possible to receive a "REMOVE" action
+      case SXE.ACTIONS.REMOVE:
+        return;
+    }
+  }
+
+  /**
+   * executes a "NEW" record
+   * @param {Object} recordProps
+   */
+  createNew(recordProps) {
+    // the value HAS to be false to ignore creation
+    // if the value is undefined or not given, proceed normally
+    if (recordProps.visible === false)
+      return;
+
+    switch (recordProps.type) {
+      case SXE.TYPES.ELEMENT: {
+        let { rid, name } = recordProps;
+        let view = this.viewManager.createView(name);
+        if (!view)
+          return;
+        view.rid = rid;
+        this.viewManager.register(view);
+        return view;
+      }
+      case SXE.TYPES.ATTR: {
+        let { parent, chdata, rid, name } = recordProps;
+        let view;
+        if (name === SVG.PROPERTIES.D) {
+          //path objects cannot be correctly displayed without the "command" or "d" parameter
+          //we have to re-create the path because fabricjs doesnt respond to dynamic changes to the path
+          let old = this.viewManager.getView(parent);
+          this.viewManager.deleteView(parent);
+          //create the correct path
+          view = this.viewManager.createView(SVG.PATH, chdata);
+          //copy all attributes
+          view.rid = old.rid;
+          for (let attr of Object.values(old.$attrs))
+            view.$createAttr(attr.name, attr.chdata, attr.rid);
+          this.viewManager.register(view);
+        } else {
+          view = this.viewManager.getView(parent);
+        }
+        if (!view)
+          return;
+        let attr = view.$createAttr(name, chdata, rid);
+        delete attr.isDirty;
+        this.canvas.$fitObject(view);
+        return attr;
+      }
+    }
+  }
+
+  /**
+   * applies a "SET" record
+   * @param {Object} recordProps
+   */
+  set(recordProps) {
+    let { chdata, visible, target } = recordProps;
+    // set visible="false" is a workaround to delete elements
+    if (visible === "false") {
```

Comment:
Why do you compare in L.66 with boolean `false` and here with string `"false"`. Did I missed a conversion step or is this JS magic I don't know? I would assume that the string version is slower...

Reply:
TLDR; *NEW* records have the value as a Boolean, while *SET* records have the value as a string.

In Java:
A NEW record represents the actual object in Java, so the value is Boolean.
SET records describe a change done on the object, so when a SET record is received while syncing, it contains the serialised value of "false" or "true".

When applying, the value is converted to Boolean, however, SET records are saved exactly the way the are received without the de-serialisation step (the idea behind saving them is to have a history of the states of the object so that in case of a conflict or sync problems, the object can be restored to a previous state).

So the conversion happens only when applying the SET records to a NEW record, resulting in Boolean value in NEW records, but remaining as a string in SET records.

This remains the same when applying the records in the HTML Whiteboard, because the records are taken "as is" from Java.

The whole idea of setting visible to false instead of using REMOVE records was a kind of "hack" to improve performance ( the SXE REMOVE record requires that the element has no children, which means we have to send REMOVE records for all of the children from the bottom up until we reach the parent, resulting in a big overhead)